pkg/controller/job/pod_failure_policy.go

```diff
@@ -97,7 +99,7 @@ func getMatchingContainerFromList(containerStatuses []v1.ContainerStatus, requir
 			continue
 		}
 		if requirement.ContainerName == nil || *requirement.ContainerName == containerStatus.Name {
-			if containerStatus.State.Terminated.ExitCode != 0 {
+			if containerStatus.State.Terminated.ExitCode != 0 || feature.DefaultFeatureGate.Enabled(features.AllowZeroInPodFailurePolicy) {
```

Comment:
Drop this line. We will make handling of `In` and `NotIn` consistent. We also don't need the feature gate.